### PR TITLE
Database Migration

### PR DESCRIPTION
**Pull request by codeknight3**
Added methods for initializing SyncClient and uploading Groups and Notifications. I am going to add more commits to this branch for retrieving server data.

### CODE REVIEW

File: src/main/java/Model/Group.java [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
     private String name;
     private String groupLink;
+    private ArrayList<String> notificationsList;
+    private ArrayList<String> pendingRequestsList;
+    private ArrayList<String> usersList;
 
+    public Group(){
+        this.id = null;
+        this.name = null;
+        this.groupLink = null;
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:16]: Do we really required to initialize them with `null`?

File: src/main/java/Model/Group.java [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
+    }
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:17]: Add a extra new line to separate out both the constructor.

File: src/main/java/Model/Group.java [hunk 3]
```diff
@@ -40,4 +50,25 @@ public String getGroupLink() {
     public void setGroupLink(String groupLink) {
         this.groupLink = groupLink;
     }
+    public void setNotificationsList(ArrayList<String> notificationsList){
+        this.notificationsList = notificationsList;
+    }
+    public void setPendingRequestsList(ArrayList<String> pendingRequestsList){
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:56]: Add a single new line

File: src/main/java/Model/Group.java [hunk 4]
```diff
@@ -40,4 +50,25 @@ public String getGroupLink() {
...
+        this.pendingRequestsList = pendingRequestsList;
+    }
+    public void setUsersList(ArrayList<String> usersList){
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:59]: add a new single line

File: src/main/java/Model/Group.java [hunk 5]
```diff
@@ -40,4 +50,25 @@ public String getGroupLink() {
...
+        this.usersList = usersList;
+    }
+
+    public ArrayList<String>getNotificationsList(){
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:63]: put a space between return type and method name

File: src/main/java/Model/Group.java [hunk 6]
```diff
@@ -40,4 +50,25 @@ public String getGroupLink() {
...
+        return this.notificationsList;
+    }
+
+    public ArrayList<String>getPendingRequestsList(){
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:67]: put a space between return type and method name

File: src/main/java/Model/Group.java [hunk 7]
```diff
@@ -40,4 +50,25 @@ public String getGroupLink() {
...
+        return this.pendingRequestsList;
+    }
+
+    public ArrayList<String>getUsersList(){
```
Comment (lakshyagupta21) [inline src/main/java/Model/Group.java:71]: put a space between return type and method name.

File: src/main/java/Model/Notification.java [hunk 8]
```diff
@@ -11,7 +12,12 @@
     private String type;
     private String status;
     private String date_str;
+    private String attachmentPath;
+    private ArrayList<String>responseList;
```
Comment (lakshyagupta21) [inline src/main/java/Model/Notification.java:16]: Add space between reference type and reference name

File: src/main/java/Data/ServerDatabaseCommunicator.java [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.*;
+
+
+public class ServerDatabaseCommunicator {
```
Comment (lakshyagupta21) [inline src/main/java/Data/ServerDatabaseCommunicator.java:21]: @DreamyPhobic @codeknight3 Can we make this class a singleton class? What are your thoughts?

**Review by lakshyagupta21 — CHANGES_REQUESTED** [review]
Not sure if this PR is up for review, but I've taken a look once and noticed couple of things where we can improve upon.

File: src/main/java/Model/Group.java (hunk 1 above)
Comment (codeknight3) [reply src/main/java/Model/Group.java]: Sorry. That was redundant. Fixed it.

**Comment by DreamyPhobic** [comment]
Convert the ServerDatabaseCommunicator class to singleton, other than that I think we are good to go. What do you say @lakshyagupta21 ??

**Review by lakshyagupta21 — APPROVED** [review]
@DreamyPhobic @codeknight3 Things looks good to me, the only thing I'm little worried about in this PR is the code quality. Anyways for now I think we can merge this and if you both also feel the same about the following issues we can take this in our next set of tasks.
For e.g.
1. variable names should start with lowercase letters.
2. Proper spacing is not being followed in the code
3. Indentations
4. Proper logging for debugging and errors (e.printStackTrace is not good way to show the error)
5. Concurrency (Right now I didn't find any such issues but when more things gets added to the project it will create problems)
6. Proper nullity checks wherever needed. (Right now everything is written assuming that we will always get the response in an expected manner, but I feel that sometimes that may not be the case because JSON response gets changed with time and for that we should have proper checks)

Solutions to fix some of the above problems
1. Integrate some code quality checker tools (e.g. checkstyle, pmd)
2. Use Annotations to avoid some runtime issues (For e.g @NonNull @Nullable)

**Comment by DreamyPhobic** [comment]
> @DreamyPhobic @codeknight3 Things looks good to me, the only thing I'm little worried about in this PR is the code quality. Anyways for now I think we can merge this and if you both also feel the same about the following issues we can take this in our next set of tasks.
> For e.g.
> 
> 1. variable names should start with lowercase letters.
> 2. Proper spacing is not being followed in the code
> 3. Indentations
> 4. Proper logging for debugging and errors (e.printStackTrace is not good way to show the error)
> 5. Concurrency (Right now I didn't find any such issues but when more things gets added to the project it will create problems)
> 6. Proper nullity checks wherever needed. (Right now everything is written assuming that we will always get the response in an expected manner, but I feel that sometimes that may not be the case because JSON response gets changed with time and for that we should have proper checks)
> 
> Solutions to fix some of the above problems
> 
> 1. Integrate some code quality checker tools (e.g. checkstyle, pmd)
> 2. Use Annotations to avoid some runtime issues (For e.g @nonnull @nullable)

Yeah I think we should have some lint and build checks on GitHub too.

**Review by lakshyagupta21 — APPROVED** [review]
Changes looks good to me, but there are still lot of places where we can do improvement not related to this PR though, but in general code needs some refactoring.

Thanks @codeknight3 for handling the code style checks so nicely. I know its very hard to fix it everywhere and that too in single PR.

**Review by jbeorse — APPROVED** [review]
Looks good to me too! Thanks for all the hard work this summer!

**Pull request merged**